04-cs-fundamentals/classroom/Finite-State-Machines.md

```diff
@@ -0,0 +1,182 @@
+
+Take a look at the state diagram, even with only 2 possible inputs at each state and a limited number of states, as the number of states increases the machine grows in complexity.  
+
+
+## A State Machine in a Table
+
+You can also represent an FSM in a table format.  The diagram and chart below illustrate the same FSM.  The left column lists the starting state and subsequent columns list the new state for the given input.  This is called a **transition table**.
+
+![Finite State Machine example 1](fsm-images/fsa1.png)
+
+
+Current State | Transition Given a 1 | Transition Given a 0
+--- | --- | ---
+A | A | B
+B | A | B
+
+
+## Exercises
+
+1. Given an alphabet of {a, b} design a FSM which will determine if the input contains `aabaa`.
+
+1. Given an alphabet of {A, C, D, G, O, T}  design a FSM which will accepts CAT or DOG alone.
+
+1. Given an alphabet of {a, b}, design an FSM which will determine if the input has an even number of `b`s.
+
+1. Given an alphabet of {BCOTY} construct an FSM which accepts CYBOT followed by any number of characters.  
+
+1. Given an alphabet of {a, b} design a FSM which will accept an only inputs with an equal number of `a`s and `b`s.
+
+1. Given an alphabet of {0, 1} design an FSM which will accept any input which does not contain `100`.
+
+1. Chris has decided to go into the smart-lock business with a keypad consisting of only 1s and 0s, along with a reset button.  He's provided the following FSM diagram to show how his lock works.
+  ![lock fsm](fsm-images/lock.png)
+
+- What is the shortest combination of keys which will open the lock?
+- Is there a longest sequence of keys which will open the lock without triggering an earlier opening?
+- Suppose Chris wanted to design a lock that required exactly 10 entries to open.  What is the minimum number of states that would be required counting the start state?
+
+8. Translate the FSM below into a table format.  You can assume things start at home.  
+  ![life Finite state machine diagram](fsm-images/life.png)
+
+9. A farm road and a highway intersect as shown below.  Detectors are placed along the farmroad to signal a vehicle is waiting to cross the highway. The traffic light controller should operate as follows. As long as no vehicle is detected on the farmroad, the lights should remain green in the highway direction. If a vehicle is detected on the farmroad, the highway lights should change from yellow to red, allowing the farmroad lights to become green. The farmroad lights stay green only as long as a vehicle is detected on the farmroad and never longer than a set interval to allow the traffic to flow along the highway. If these conditions are met, the farmroad lights change from green to yellow to red, allowing the highway lights to return to green. Even if vehicles are waiting to cross the highway, the highway should remain green for a set interval.  Design a FSM for this problem.
```

Comment:
```suggestion
9. A farm road and a highway intersect as shown below.  Detectors are placed along the farmroad to signal a vehicle is waiting to cross the highway. The traffic light controller should operate as follows. As long as no vehicle is detected on the farmroad, the lights should remain green in the highway direction. If a vehicle is detected on the farmroad, the highway lights should change from green to yellow to red, allowing the farmroad lights to become green. The farmroad lights stay green only as long as a vehicle is detected on the farmroad and never longer than a set interval to allow the traffic to flow along the highway. If these conditions are met, the farmroad lights change from green to yellow to red, allowing the highway lights to return to green. Even if vehicles are waiting to cross the highway, the highway should remain green for a set interval.  Design an FSM for this problem.
```

Reply:
Consider breaking this into smaller paragraphs.  It's a little hard to parse for humans and apparently Github just gives up on displaying a diff in the suggestion.